tests/test_config.py

```diff
@@ -384,8 +384,14 @@ def mock_getuid():
 
             return 123
 
+        def mock_home():
+            """Return a static local home directory"""
+
+            return '/home/user'
+
         with patch('os.getuid', mock_getuid):
-            config = self._parse_config('RemoteCommand %i')
+            with patch('pathlib.Path.home', mock_home):
```

Comment:
Are you actually seeing an error when pathlib.Path.home() is called? If so, on which platform? The reason this was mocked above was to control the specific value which is returned, but here I wouldn't expect that to be needed as long as the call doesn't error out, since this test doesn't actually use the value.

Looking at the "ntpath" code, I guess it could return an error if neither USERPROFILE nor HOMEPATH are set. This could be a problem for more than just the test code, though. I think a better fix here would be to switch from calling Path.home() to calling os.path.expanduser() in SSHClientConfig._set_tokens(), which will just leave the '~' unexpanded but not raise an error.

Reply:
This test is only executed on Unix, and I've seen this on both macOS and Linux. `Path.home()` raises an exception if `~` cannot be expanded. I think the reason why it does not always happen, is because on some system there is actually a user with id 123.

Reply:
Would you prefer changing `SSHClientConfig._set_tokens()` instead?

Reply:
On Linux, it seems to require a combination of having HOME unset in the environment and the UID returned by os.getuid() not existing, but you're right that it can happen there as well as on Windows.

I think changing to `os.path.expanduser('~')` is safer. Otherwise, I think the code would need to only raise an error when '%d' was actually used as a substitution which would be a much more complicated change, or always catch and ignore it, which is messier than just using expanduser which already works that way. I'm thinking of something like:

```patch
diff --git a/asyncssh/config.py b/asyncssh/config.py
index 24cbcbd..e3f37d3 100644
--- a/asyncssh/config.py
+++ b/asyncssh/config.py
@@ -477,7 +477,7 @@ class SSHClientConfig(SSHConfig):
         conn_hash = sha1(conn_info.encode('utf-8')).hexdigest()

         self._tokens.update({'C': conn_hash,
-                             'd': str(Path.home()),
+                             'd': str(os.path.expanduser('~')),
                              'h': host,
                              'L': short_local_host,
                              'l': local_host,
diff --git a/tests/test_config.py b/tests/test_config.py
index f5da922..2388a48 100644
--- a/tests/test_config.py
+++ b/tests/test_config.py
@@ -357,13 +357,13 @@ class _TestClientConfig(_TestConfig):

             return 'thishost.local'

-        def mock_home():
+        def mock_expanduser(path):
             """Return a static local home directory"""

             return '/home/user'

         with patch('socket.gethostname', mock_gethostname):
-            with patch('pathlib.Path.home', mock_home):
+            with patch('os.path.expanduser', mock_expanduser):
                 config = self._parse_config(
                     'Hostname newhost\n'
                     'User newuser\n'
```

Reply:
The mock isn't needed if we use os.path.expanduser, is it?

Reply:
Also we probably should raise an error when `%d` is used but `expanduser` fails, otherwise it expands to a literal `~` which can be confusing if it happens.

Reply:
The mock is needed for the test_percent_expansion() test, so that the %d expands to a known value the test can compare against. The one you added in test_uid_percent_expansion is not needed with this change, though.

Regarding checking if expanduser fails when %d is used, I guess a check could be added in `_expand_val()`. I'm thinking it might be best not to add '%d' to the _tokens dictionary at all in `set_tokens()` in that case (like it does when adding '%i' only when a UID is available. The exception message could possibly be customized for those two cases, but that could be done inside the handling of `KeyError`, which I think would be cleaner.

For now, I'd suggest just making the above changes in this PR, or perhaps backing out the test change which added the mock and I can take care of the change to use expanduser myself in a separate commit.

Reply:
Done.

Reply:
Looks good - thanks!